translib/db/cvl_db_access.go

```diff
@@ -0,0 +1,535 @@
+////////////////////////////////////////////////////////////////////////////////
+//                                                                            //
+//  Copyright 2023 Broadcom. The term Broadcom refers to Broadcom Inc. and/or //
+//  its subsidiaries.                                                         //
+//                                                                            //
+//  Licensed under the Apache License, Version 2.0 (the "License");           //
+//  you may not use this file except in compliance with the License.          //
+//  You may obtain a copy of the License at                                   //
+//                                                                            //
+//     http://www.apache.org/licenses/LICENSE-2.0                             //
+//                                                                            //
+//  Unless required by applicable law or agreed to in writing, software       //
+//  distributed under the License is distributed on an "AS IS" BASIS,         //
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  //
+//  See the License for the specific language governing permissions and       //
+//  limitations under the License.                                            //
+//                                                                            //
+////////////////////////////////////////////////////////////////////////////////
+
+package db
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/Azure/sonic-mgmt-common/cvl"
+	ctypes "github.com/Azure/sonic-mgmt-common/cvl/common"
+	"github.com/Azure/sonic-mgmt-common/translib/tlerr"
+	"github.com/go-redis/redis/v7"
+	log "github.com/golang/glog"
+)
+
+type cvlDBAccess struct {
+	Db *DB
+}
+
+func (d *DB) NewValidationSession() (*cvl.CVL, error) {
+	if d == nil || d.Opts.DBNo != ConfigDB {
+		return nil, tlerr.TranslibDBNotSupported{}
+	}
+
+	//TODO: INTG_CHANGES: uncomment the below commented line
+	//if c, status := cvl.ValidationSessOpen(&cvlDBAccess{d}); status != cvl.CVL_SUCCESS {
+	if c, status := cvl.ValidationSessOpen(); status != cvl.CVL_SUCCESS {
+		return nil, tlerr.TranslibCVLFailure{Code: int(status)}
+	} else {
+		return c, nil
+	}
+}
+
+func NewValidationSession() (*cvl.CVL, error) {
+	//TODO: INTG_CHANGES: uncomment the below commented line
+	//if c, status := cvl.ValidationSessOpen(nil); status != cvl.CVL_SUCCESS {
+	if c, status := cvl.ValidationSessOpen(); status != cvl.CVL_SUCCESS {
+		return nil, tlerr.TranslibCVLFailure{Code: int(status)}
+	} else {
+		return c, nil
+	}
+}
+
+func (c *cvlDBAccess) Exists(key string) ctypes.IntResult {
+	keys, err := c.Keys(key).Result()
+	switch err.(type) {
+	case tlerr.TranslibRedisClientEntryNotExist:
+		err = redis.Nil
+	}
+	if len(keys) > 1 {
+		//TODO have an optimized implementation in DBAL for Exists
+		return intResult{int64(0), err}
```

Comment:
Why do we return 0 if keys exist? Also, would it make more sense to return a bool instead of an int?

Reply:
These APIs are wrapper for proper redis APIs hence the signature is same. https://redis.io/commands/exists/

Reply:
excepting the single key match, since the count is more than one, it is match for pattern not for single key, hence returning 0